Add `ignore` from Zygote to ChainRulesCore (#229)
JuliaDiff/ChainRulesCore.jl: docs/src/tips_for_packages.md

# Tips for making your package work with AD

## Ignoring gradients for certain expressions

There exists code that is not meant to be differentiated through, for example logging.
In some cases, AD systems might work perfectly well with that code, but in others they might not.
A convenience function `ignore_gradients` is provided to get around this issue.

> It captures the functionality of both `Zygote.ignore` and `Zygote.dropgrad`

Reviewer: ```suggestion
It captures the functionality of both `Zygote.ignore` and `Zygote.dropgrad`.
```

> For example, Zygote will break if you try to store intermediate values like so because it does not support mutation.

Reviewer: ```suggestion
For example, Zygote does not support mutation, so it will break if you try to store intermediate values as in the following example:
```

```julia
somes = []
things = []

function loss(x, y)
    some = f(x, y)
    thing = g(x)
    
    # log
    push!(somes, some)
    push!(things, thing)

    return some + thing
end
```

It is possible to get around this by using the `ignore_gradients` function
```julia
ignore_gradients() do
    push!(somes, some)
    push!(things, thing)
end
```
or using a macro for one-liners
```julia
@ignore_gradients push!(things, thing)
```

It is also possible to use this on individual objects, e.g.
```julia
ignore_gradients(a) + b
```

> will ignore the gradients for `a` only.

Reviewer: Thanks for adding this - really clear and helpful! Could you also include a comment that highlights the difference in behaviour between functions (they get called, behavior ignored) and functors (they don't get called, but gradients w.r.t to their fields will get ignored? though I'm not even sure on this!)
